pennylane/operation.py

```diff
@@ -528,6 +529,123 @@ def __init__(self, *args, wires=None, do_queue=True):
 
         super().__init__(*params, wires=wires, do_queue=do_queue)
 
+    def __matmul__(self, other):
+        if isinstance(other, Tensor):
+            return other.__rmatmul__(self)
+
+        if isinstance(other, Observable):
+            return Tensor(self, other)
+
+        raise ValueError("Can only perform tensor products between observables.")
+
+
+class Tensor(Observable):
+    """Container class representing tensor products of observables.
+
+    To create a tensor, simply initiate it like so:
+
+    >>> T = Tensor(qml.PauliX(0), qml.Hermitian(A, [1, 2]))
+
+    You can also create a tensor from other Tensors:
+
+    >>> T = Tensor(T, qml.PauliZ(4))
+
+    The ``@`` symbol can be used as a tensor product operation:
+
+    >>> T = qml.PauliX(0) @ qml.Hadamard(2)
+    """
+    return_type = None
+    tensor = True
+    par_domain = None
+
+    def __init__(self, *args): #pylint: disable=super-init-not-called
+        self.obs = []
+
+        for o in args:
+            if isinstance(o, Tensor):
+                self.obs.extend(o.obs)
+            elif isinstance(o, Observable):
+                self.obs.append(o)
+            else:
+                raise ValueError("Can only perform tensor products between observables.")
+
+    def __str__(self):
+        """Print the tensor product and some information."""
+        return 'Tensor product {}: {} params, wires {}'.format([i.name for i in self.obs], len(self.params), self.wires)
+
+    @property
+    def name(self):
+        """All consituent observable names making up the tensor product.
+
+        Returns:
+            list[str]: list containing all observable names
+        """
+        return [o.name for o in self.obs]
+
+    @property
+    def num_wires(self):
+        """Number of wires the tensor product acts on.
+
+        Returns:
+            int: number of wires
+        """
+        return len([w for sublist in [o.wires for o in self.obs] for w in sublist])
+
+    @property
+    def wires(self):
+        """All wires in the system the tensor product acts on.
+
+        Returns:
+            list[list[Any]]: nested list containing the wires per observable
+                in the tensor product
+        """
+        return [o.wires for o in self.obs]
+
+    @property
+    def params(self):
+        """Raw parameters of all constituent observables in the tensor product.
+
+        Returns:
+            list[Any]: flattened list containing all dependent parameters
+        """
+        return [p for sublist in [o.params for o in self.obs] for p in sublist]
```

Comment:
Can we not use `self.parameters` here?
What is the difference between `o.parameters` (used in `def parameters` below) and `o.params` used here? From what I can see, attempting to access `o.params` will raise an exception

Reply:
Actually, I find having two functions `self.params` and `self.parameters` to be confusing

Reply:
This is an artifact of the `Tensor` class needing to satisfy the `Operation` class API, and valid `Operations` need to have both defined :slightly_frowning_face:  Not ideal.

In the `Operation` class, `.params` is used for the unevaluated parameters (so could contain `Variable` instances), while `.parameters` are evaluated.

The QNode always queries `Operation.params`, while the device always queries `Operation.parameters`

Reply:
Something for a quick refactor (name change would be sufficient, e.g., `evaluated_params`)?